apps/emqx_authn/test/emqx_authn_mongo_SUITE.erl

```diff
@@ -0,0 +1,409 @@
+
+    Credentials = Credentials0#{
+                    listener => 'tcp:default',
+                    protocol => mqtt
+                   },
+    ?assertEqual(Result, emqx_access_control:authenticate(Credentials)),
+
+    emqx_authn_test_lib:delete_authenticators(
+      [authentication],
+      ?GLOBAL).
+
+t_destroy(_Config) ->
+    ok = init_seeds(),
+    AuthConfig = raw_mongo_auth_config(),
+
+    {ok, _} = emqx:update_config(
+                ?PATH,
+                {create_authenticator, ?GLOBAL, AuthConfig}),
+
+    {ok, [#{provider := emqx_authn_mongodb, state := State}]}
+        = emqx_authentication:list_authenticators(?GLOBAL),
+
+    {ok, _} = emqx_authn_mongodb:authenticate(
+                #{username => <<"plain">>,
+                  password => <<"plain">>
+                 },
+                State),
+
+    emqx_authn_test_lib:delete_authenticators(
+      [authentication],
+      ?GLOBAL),
+
+    % Authenticator should not be usable anymore
+    ?assertException(
+       error,
+       _,
+       emqx_authn_mongodb:authenticate(
+         #{username => <<"plain">>,
+           password => <<"plain">>
+          },
+         State)),
+
+    ok = drop_seeds().
+
+t_update(_Config) ->
+    ok = init_seeds(),
+    CorrectConfig = raw_mongo_auth_config(),
+    IncorrectConfig =
+        CorrectConfig#{selector => #{<<"wrongfield">> => <<"wrongvalue">>}},
+
+    {ok, _} = emqx:update_config(
+                ?PATH,
+                {create_authenticator, ?GLOBAL, IncorrectConfig}),
+
+    {error, not_authorized} = emqx_access_control:authenticate(
+                                #{username => <<"plain">>,
+                                  password => <<"plain">>,
+                                  listener => 'tcp:default',
+                                  protocol => mqtt
+                                 }),
+
+    % We update with config with correct selector, provider should update and work properly
+    {ok, _} = emqx:update_config(
+                ?PATH,
+                {update_authenticator, ?GLOBAL, <<"password-based:mongodb">>, CorrectConfig}),
+
+    {ok,_} = emqx_access_control:authenticate(
+               #{username => <<"plain">>,
+                 password => <<"plain">>,
+                 listener => 'tcp:default',
+                 protocol => mqtt
+                }),
+    ok = drop_seeds().
+
+t_is_superuser(_Config) ->
+    Config = raw_mongo_auth_config(),
+    {ok, _} = emqx:update_config(
+                ?PATH,
+                {create_authenticator, ?GLOBAL, Config}),
+
+    Checks = [
+              {<<"0">>,     false},
+              {<<"">>,      false},
+              {null,        false},
+              {false,       false},
+              {0,           false},
+
+              {<<"1">>,     true},
+              {<<"val">>,   true},
+              {1,           true},
+              {123,         true},
+              {true,        true}
+             ],
```

Comment:
Maybe test the new `undefined` case?

Reply:
It is caught by this test, we get undefined when insert null atom :)